File: js/pigLatin.js
translate = (sentence) => {
  return sentence.split(" ").map(word => translateWord(word)).join(" ");
};

translateWord = (word) => {
  let output = "";
  let punctuation = "";
  const reVowelWords = /^[aeiou]/
  const reTranslateQuWords = /^.*qu/
  const reConsonantWords = /^[^aeiou]*/
  const rePunctuation = /[^\w][A-Za-z]*$/
  // remember if the word is upper case
  let wasUpper = isUpper(word)
  // Set word to lower case
  word = word.toLowerCase()
  // set aside punctuation, remove it from the string
  if (rePunctuation.test(word)) {
    punctuation = word.match(rePunctuation)[0]
  }
  word = word.replace(rePunctuation, "")
  // translate the words now...
  if (reVowelWords.test(word)) {
    output = `${word}ay`
  } else if (reTranslateQuWords.test(word)) {
    output = rearrange(word, reTranslateQuWords)
  } else if (reConsonantWords.test(word)) {
    output = rearrange(word, reConsonantWords)
  }
  // re apply any punctuation from earlier
  output += punctuation;
  // return a correctly cased word
  return wasUpper ? capitalize(output) : output;
}

rearrange = (word, re) => {
  let match = word.match(re)
  return word.slice(match[0].length) + match[0] + "ay"
}

isUpper = (word) => {
  let firstCharCode = word.charCodeAt(0);
  return firstCharCode > 65 && firstCharCode < 90;
}

capitalize = (word) => {
  return word[0].toUpperCase() + word.slice(1)
}

module.exports = {
  translate
}

console.log(translate("Completed the challenge!"))

File: python/pig_latin.py
"""
translate a phrase to pig-latin.
1. Translate
2. Keep punctuation
3. Keep capitalization
"""
import re

def translate(phrase):
    """
    splits phrase into a list, sends each word to translate word
    """
    return " ".join([translate_word(word) for word in phrase.split()])

def is_vowel_word(word):
    """
    Returns the match object from looking for a vowel at the beginning of the word
    """
    re_vowel_words = re.compile(r"^[aeiou]")
    return re_vowel_words.search(word)

def is_qu_word(word):
    """
    returns the match object from finding a 'qu' at the beginning of a word
    """
    re_translate_qu_words = re.compile(r"^.*qu")
    return re_translate_qu_words.search(word)

def is_constonant_word(word):
    """
    returns the match object from finding constonants at the beginning of a word
    """
    re_consonant_words = re.compile(r"^[^aeiou]*")
    return re_consonant_words.search(word)

def translate_word(word):
    """
    split a phrase in to multiple words
    submit that to a word translator
    """
    # set up output
    output = ""
    # remember punctuation
    punctuation = ""
    re_punctuation = re.compile(r"[^\w]*$")
    # remember if upper
    was_upper = is_upper(word)
    # move to lowercase
    word = word.lower()
    # set aside punc, remove it from the string
    punctuation = re_punctuation.search(word).group(0)
    word = re_punctuation.sub("", word)
    # translate the words with regexes
    if is_vowel_word(word):
        output = f"{word}ay"
    elif is_qu_word(word):
        match = is_qu_word(word).group(0)
        output = word[len(match):] + match + "ay"
    elif is_constonant_word(word):
        match = is_constonant_word(word).group(0)
        output = word[len(match):] + match + "ay"
    # put the word/puncuation back together
    output += punctuation
    # return the capitalized if necessary word
    return capitalize(output) if was_upper else output

def is_upper(word):
    """
    checks if the only the first character of a word is uppercase
    """
    return word[0].isupper()

def capitalize(word):
    """
    returns the word with the first letter capitalized
    """
    return word[0].upper() + word[1:]
